### DISCUSSION TITLE
Serialize raw SQL fragment 'null'

### DISCUSSION THREAD

**Discussion by Darent98** [mikro-orm/mikro-orm#5257] [Q&A] [answered]
### Describe the bug

I discovered an issue regarding the serialization of RawQueryFragment when the SQL query is set to 'null'. Neither the toJSON method nor JSON.stringify are functioning properly in this scenario.

### Reproduction

``` ts
import { RawQueryFragment, raw } from '@mikro-orm/sqlite';

test('serialize RawQueryFragment - null', async () => {
  const tmp: RawQueryFragment = raw('null')
  // expected QueryFragment can be serialized
  expect(() => tmp.toJSON()).not.toThrow()
});
```

### What driver are you using?

None

### MikroORM version

6.1.5-dev.0

### Node.js version

v20.10.0

### Operating system

Linux

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
This is expected behavior, why would you want to serialize it? It is there to protect you from ending up with raw statements instead of values when you serialize entity.

Note that this has nothing to do with the value, it works the same with `raw('1')` or anything else, its not about the `null`.

**Comment by Darent98** [comment]
The problem occurs, when I am trying to log a working insert query, which happens when I run the app in debug mode.
In the executeQuery method in the Connection.js file the context consists of params and a query. In the params I got a RawQueryFragment, which correctly gets insert into the query at the '?'. The query runs as expected but trying to log this, throws the exception. 
Probably the right behavior would be to disallow this completely or to allow the serialization of the RawSqlFragment.

**Reply by B4nan** [reply]
Show some code, doing `console.log(qb)` works fine, if you want to log the query, you should be using `qb.getFormattedQuery()` or other methods, not serializing the QB instance. Alternatively, use `util.inspect` which will also work fine.

> Probably the right behavior would be to disallow this completely

Disallow what?

> or to allow the serialization of the RawSqlFragment.

Let me repeat that this is intentional.

**Reply by Darent98** [reply]
Thanks for your response, with your provided information I probably can workaround my problem.
To give a better context: I want to send the complete context to Loki (params needed for potential debugging). For that the context gets serialized. 
``` ts
logQuery (context: { query: string } & LogContext): void {
    return getLogger('query').debug?.(context.took === undefined ? ':query' : ':took ms - :query', context)
  }
```
context.params containing: 
![image](https://github.com/mikro-orm/mikro-orm/assets/84193036/5e5cd532-8567-4be3-b108-18cff81e5949)

The workaround could be to check if params containing a RawQueryFragment, returning the sql and also checking if the param contains further params to run this recursively.

It would be nice if RawSqlFragments like this could be serzialized, but I have no idea how to achive this while keeping the protection of running into raw statements instead of values.

Thanks for your help and I hope I could clarify why this error occured in the first place.

**Reply by B4nan** [reply]
I see, maybe we could have that validation enabled only when serializing some entity, since this operation is synchronous, we can use a simple static boolean switch on the class itself and toggle it from inside EntityTransformer, that could be actually pretty simple and it would resolve this problem, its fine to serialize the value itself when its not in the context of serializing entity.

**Reply by Darent98** [reply]
That would be a neat enhancement, even better that it can be implemented fairly simple.

**Reply by B4nan** [reply, marked as answer]
https://github.com/mikro-orm/mikro-orm/commit/9158f515e513310f593aade04e6bedaa29cd6459?w=1
